engine/runtime/src/main/java/org/enso/interpreter/node/callable/argument/ReadArgumentCheckNode.java

```diff
@@ -0,0 +1,175 @@
+import org.enso.interpreter.runtime.data.Type;
+import org.enso.interpreter.runtime.error.DataflowError;
+import org.enso.interpreter.runtime.error.PanicException;
+import org.graalvm.collections.Pair;
+
+public abstract class ReadArgumentCheckNode extends Node {
+  private final String name;
+  @Child IsValueOfTypeNode checkType;
+  @CompilerDirectives.CompilationFinal(dimensions = 1)
+  private final Type[] expectedTypes;
+  @CompilerDirectives.CompilationFinal
+  private String expectedTypeMessage;
+
+  ReadArgumentCheckNode(String name, Type[] expectedTypes) {
+    this.name = name;
+    this.checkType = IsValueOfTypeNode.build();
+    this.expectedTypes = expectedTypes;
+  }
+
+  static ReadArgumentCheckNode build(String argumentName, Type[] expectedTypes) {
+    if (expectedTypes == null || expectedTypes.length == 0) {
+      return null;
+    } else {
+      return ReadArgumentCheckNodeGen.create(argumentName, expectedTypes);
+    }
+  }
+
+  abstract Object executeCheckOrConversion(VirtualFrame frame, Object value);
+
+  @Specialization(rewriteOn = InvalidAssumptionException.class)
+  Object doCheckNoConversionNeeded(VirtualFrame frame, Object v) throws InvalidAssumptionException {
+    if (findAmongTypes(v)) {
+      return v;
+    } else {
+      throw new InvalidAssumptionException();
+    }
+  }
+
+  @Specialization(limit = "10", guards = {
+      "cachedType != null",
+      "findType(typeOfNode, v) == cachedType"
+  })
+  Object doWithConversionCached(
+    VirtualFrame frame, Object v,
+    @Cached TypeOfNode typeOfNode,
+    @Cached("findType(typeOfNode, v)") Type cachedType,
+    @Cached("findConversionNode(cachedType)") ApplicationNode convertNode
+  ) {
+    return handleWithConversion(frame, v, convertNode);
+  }
+
+  @Specialization(replaces = "doWithConversionCached")
+  Object doWithConversionUncached(
+    VirtualFrame frame, Object v,
+    @Cached TypeOfNode typeOfNode
+  ) {
+    var type = findType(typeOfNode, v);
+    return doWithConversionUncachedBoundary(frame.materialize(), v, type);
+  }
+
+  private static boolean isAllFitValue(Object v) {
+    return v instanceof DataflowError
+            || (v instanceof Function fn && fn.isThunk());
+  }
+
+  @ExplodeLoop
+  private boolean findAmongTypes(Object v) {
+    for (Type t : expectedTypes) {
+      if (checkType.execute(t, v)) {
+        return true;
+      }
+    }
+    return isAllFitValue(v);
+  }
+
+  @ExplodeLoop
+  private Pair<Function, Type> findConversion(Type from) {
+    var ctx = EnsoContext.get(this);
+
+    if (getRootNode() instanceof EnsoRootNode root) {
+      var convert = UnresolvedConversion.build(root.getModuleScope());
+      for (Type into : expectedTypes) {
+        var conv = convert.resolveFor(ctx, into, from);
+        if (conv != null) {
+          return Pair.create(conv, into);
+        }
+      }
+    }
+    return null;
+  }
+
+  ApplicationNode findConversionNode(Type from) {
+    var convAndType = findConversion(from);
```

Comment:
`from` can be `null` I think, but I don't think we handle that nicely in `findConversion`, do we?